### ISSUE TITLE
trino-server-rpm doesn't support CentOS/RHEL 8 due to lack of unversioned python package

### ISSUE THREAD

**Issue by sumit-gupta-sgt** [trinodb/trino#10743]
When I am trying to install trino-server-rpm-368.rpm over rhel8 it always says python >= 2.4, i have python2.7 and python3.6 in my system already.

While i was using same package over centos7..everything works well.

**Comment by mosabua** [comment]
You need to ensure that the desired Python version is on the path for the user that you use to start the rpm with.

**Comment by hashhar** [comment]
@sumit-gupta-sgt I believe you are having errors installing the RPM itself?
Can you share exactly what error RPM seems to throw when you install the rpm?

Also can you share output of `rpm -qa | grep python`?

Most probably RHEL8 has removed the `python` package and replaced it with explicit `python2` and `python3` packages which cause the RPM to error out since it cannot find a `python` package.

As a workaround you can still try to install the rpm by doing something like `rpm -i --no-deps <your_rpm_file>` which would ignore all listed dependencies and still work if you have the dependencies actually installed.

**Comment by sumit-gupta-sgt** [comment]
@hashhar @mosabua mosabuaI am using same version of python in centos7, everything work well. When i try to use trino rpm, it always says python >= 2.4 required, while i have python 2.7 and python3 in my system.

**Comment by hashhar** [comment]
The issue is that CentOS 8 doesn't have any package that *provides* `python` - the packages now either provide `python2` or `python3` so the RPM installation fails since we request `python >= 2.4` which cannot be satisfied.

The only workaround at this time is what I shared above:

```
yum install python2
rpm -i --nodeps <your_rpm_file>
```

I'm looking into ways to build a single RPM that would work on both RHEL 8 based as well as older systems.

**Comment by hashhar** [comment]
Seems like it's impossible to support both CentOS >= 8 as well as older versions with a single RPM file.

We'd need to create and distribute two different RPM files for each packaging variation. The gist is that the dependencies declared in RPM are static and cannot be conditional so we cannot alter behaviour of RPM according to what distro we are running on.

What we can do is to remove `python` as a dependency and provide a pre-install script that detects lack of an unversioned python executable and fails with some instructions - similar to how we do JDK detection today.

cc: @martint Any opinions? I'd prefer the similar solution that we have for JDK detection and remove `python` as a declared package dependency.

**Comment by martint** [comment]
That seems reasonable. Although, if CentOS < 8 is old and no longer supported, we might want to drop support for it altogether. If someone wants to install it in that distro, they can bypass the dependency check.

**Comment by sumit-gupta-sgt** [comment]
@hashhar When i try to use workaround you mentioned, looks like didn't work.

rpm -i --nodeps trino-server-rpm-368.rpm
package trino-server-rpm-0:368-1.noarch does not verify: no digest

Please let me know if anything i am missing.

**Comment by hashhar** [comment]
Seems like you have FIPS enabled on your CentOS installation. You can skip the digest checking via `--nodigest --nofiledigest` options passed to `rpm`. Let's keep this issue focused on building CentOS/RHEL 8 compatible RPM.

You can join our [Slack](https://trino.io/slack.html) and continue the conversation there if you still run into issues.

**Comment by sumit-gupta-sgt** [comment]
@hashhar thanks let me try.

BTW when you are planning to have separate RPM ..any plan schedule ?

**Comment by mosabua** [comment]
I agree that it might be easier to have a script that checks for suitable python on the path. This could even be used in the tarball as well. 

And on that note of tarball .. @sumit-gupta-sgt you can just use the tarball on Centos .. you just have to manually satisfy dependencies and hook up the launcher script as daemon (only if desired..)

**Comment by sumit-gupta-sgt** [comment]
@hashhar Quick note on command provide by you.

yum install python2
sudo alternatives --set python /usr/bin/python2
rpm -i --nodeps <your_rpm_file>

Lets assume i don't have python2 in my system and when i try to do sudo alternatives --set python /usr/bin/python3, Trino still says python >= 2.4 needed by trino-server-rpm.

**Comment by sumit-gupta-sgt** [comment]
Any Update here ?

This is blocker if you don't have python2 on RHEL8.

**Comment by ellieshen** [comment]
@hashhar , any plan to remove python in RPM as a dependency to compatible CentOS/RHEL 8?

**Comment by hashhar** [comment]
Python is REQUIRED for the launcher to work. At this point we either need separate RPMs per OS version (which is not something we want to do) or have some hacky way of identifying whether expected Python version is available or not (which may or may not work on all distributions and still leaves the user to install proper Python version themselves).

These are some of the reasons why we have considered dropping the RPM as a distribution option in the past. The tarball can be used on all OSes and doesn't have any of these problems.
